jupyter/sum-to-zero/sum_to_zero_evaluation.qmd

```diff
@@ -0,0 +1,1582 @@
+
+In order to do this multiway comparison, we assemble the individual summaries from the 6 runs above
+into two dataframes.  We also compute the number of effective samples per second - a key metric of
+model efficiency.
+(Note:  we're using a development version of CmdStanPy to scrape time information from the CSV files
+because changes to CmdStan's `stansummary` function removed the ESS/sec metric.
+This is a workaround for now).
+
+```{python}
+# small dataset
+ozs_fit_summary = binomial_ozs_fit.summary(sig_figs=2)
+ozs_fit_summary.index =  ozs_fit_summary.index.astype(str) + "  a) ozs"
+ozs_fit_time = binomial_ozs_fit.time
+ozs_total_time = 0
+for i in range(len(ozs_fit_time)):
+    ozs_total_time += ozs_fit_time[i]['total']
+ozs_fit_summary['ESS/sec'] = ozs_fit_summary['ESS_bulk']/ozs_total_time
+
+hard_fit_summary = binomial_hard_fit.summary(sig_figs=2)
+hard_fit_summary.index = hard_fit_summary.index.astype(str) + "  b) hard"
+hard_fit_time = binomial_hard_fit.time
+hard_total_time = 0
+for i in range(len(hard_fit_time)):
+    hard_total_time += hard_fit_time[i]['total']
+hard_fit_summary['ESS/sec'] = hard_fit_summary['ESS_bulk']/hard_total_time
+
+soft_fit_summary = binomial_soft_fit.summary(sig_figs=2)
+soft_fit_summary.index = soft_fit_summary.index.astype(str) + "  c) soft"
+soft_fit_time = binomial_soft_fit.time
+soft_total_time = 0
+for i in range(len(soft_fit_time)):
+    soft_total_time += soft_fit_time[i]['total']
+soft_fit_summary['ESS/sec'] = soft_fit_summary['ESS_bulk']/soft_total_time
+
+small_data_fits_summary = pd.concat([ozs_fit_summary, hard_fit_summary, soft_fit_summary])
+
+# large dataset
+ozs_fit_lg_summary = binomial_ozs_fit_lg.summary(sig_figs=2)
+ozs_fit_lg_summary.index =  ozs_fit_lg_summary.index.astype(str) + "  a) ozs"
+ozs_fit_time_lg = binomial_ozs_fit_lg.time
+ozs_total_time_lg = 0
+for i in range(len(ozs_fit_time_lg)):
+    ozs_total_time_lg += ozs_fit_time_lg[i]['total']
+ozs_fit_lg_summary['ESS/sec'] = ozs_fit_lg_summary['ESS_bulk']/ozs_total_time_lg
+
+hard_fit_lg_summary = binomial_hard_fit_lg.summary(sig_figs=2)
+hard_fit_lg_summary.index = hard_fit_lg_summary.index.astype(str) + "  b) hard"
+hard_fit_time_lg = binomial_hard_fit_lg.time
+hard_total_time_lg = 0
+for i in range(len(hard_fit_time_lg)):
+    hard_total_time_lg += hard_fit_time_lg[i]['total']
+hard_fit_lg_summary['ESS/sec'] = hard_fit_lg_summary['ESS_bulk']/hard_total_time_lg
+
+soft_fit_lg_summary = binomial_soft_fit_lg.summary(sig_figs=2)
+soft_fit_lg_summary.index = soft_fit_lg_summary.index.astype(str) + "  c) soft"
+soft_fit_time_lg = binomial_soft_fit_lg.time
+soft_total_time_lg = 0
+for i in range(len(soft_fit_time_lg)):
+    soft_total_time_lg += soft_fit_time_lg[i]['total']
+soft_fit_lg_summary['ESS/sec'] = soft_fit_lg_summary['ESS_bulk']/soft_total_time_lg
+
+large_data_fits_summary = pd.concat([ozs_fit_lg_summary, hard_fit_lg_summary, soft_fit_lg_summary])
+```
+
+**Eth**
+
```

Comment:
What is this? What does all the code do? Don't assume your readers know python that well!

Reply:
@spinkney: What are you asking for here?  Documentation within the code?  A description summarizing what it does on the outside?  

I would assume the readers know Python pretty well.  Otherwise, the intro to Python will overwhelm the intro to sum-to-zero.  If you do want to teach people bits of Python, you should have a model reader in mind or it will be very hard to decide where to stop with expounding on the Python code.  Is that someone who knows Python syntax but not any libraries like numpy or matplotlib?  Or someone even less skilled who doesn't understand python scoping or control flow?  Maybe they don't know what `+=` means or that `range(N)` is from 0 to `N-1`?